docs/allocation.md

```diff
@@ -0,0 +1,152 @@
+
+The garbage collector reclaims memory by removing objects when they are no longer required. To find out more about the garbage collector, see [Garbage collection](gc_overview.md).
+
+Depending on your application workload or service level agreement, you can choose from a number of OpenJ9 *garbage collection (GC) policies*. Each GC policy uses a different strategy to manage memory on the heap. The structure of the heap also depends on the strategy in force. For more information about choosing a GC policy, see [Garbage collection policies](gc.md).
+
+## The allocator
+
+The allocator manages pools of free memory and how the free memory is consumed. It is also responsible for allocating areas of storage in the Java heap for objects at the request of applications, class libraries, or the VM.
+
+Every allocation requires a *heap lock* to stop concurrent threads trying to access the same area of memory at the same time. When an object is allocated, the heap lock is released. If there is insufficient space to allocate the object, allocation fails, the heap lock is released, and the GC is called. If the GC manages to recover some space on the heap, the allocator can resume operations. If the GC does not recover enough space, it returns an `OutOfMemoryError` exception.
+
+Acquiring a heap lock for every allocation would be an intensive operation with a knock on impact to performance. To get around this problem, small objects are allocated to allocation caches.
+
+### Allocation caches
+
+To improve performance, allocation caches are reserved in the heap for different threads. These allocation caches are known as thread local heaps (TLH) and allow each thread to allocate memory from its cache without acquiring the heap lock. A TLH is typically used for small objects of less than 512 bytes (768 bytes on 64-bit VMs) although larger objects can be allocated from the cache if there is sufficient space.
+
+If a thread allocates a lot of objects, the allocator gives that thread a larger TLH to reduce contention on the heap lock.
+
+A TLH is predefined with an initial default size of 2 KB. On every TLH refresh, the requested size for that thread
+is increased by an increment (default 4 KB). The requested size can grow up to a predefined maximum (default 128 KB).
+
+After every garbage collection cycle, the TLH requested size for each thread is reduced, sometimes by as much as 50%, to take account of threads that
+reduce their allocation rate and no longer need large TLHs.
+
+For very inactive threads, the requested size can even drop below the initial value, down to the predefined minimum (512/768 bytes).
+For very active threads, the maximum TLH requested size might be reached before the next GC occurs.
+
+Larger TLHs can help reduce heap lock contention, but might also reduce heap utilization and increase heap fragmentation.
```

Comment:
I don't see it explicitly stated, since it may be obvious and or implied from previous section, but I'll mention it anyway:

When TLH refresh is done, heap lock is used.

Failing to refresh TLH is treated the same as failing to satisfy an individual allocation: it triggers a GC.

Reply:
Updated the following a couple of paragraphs up:

"If a TLH refresh fails to complete, a GC cycle is triggered."